### DISCUSSION TITLE
LAN865x crashes on ESP32C3 and ESP32S3

### DISCUSSION THREAD

**Discussion by lukas2511** [zephyrproject-rtos/zephyr#78021] [Q&A] [answered]
I'm currently trying to get 10BASE-T1S connectivity on an ESP32 based platform. I've tried both ESP32C3 and ESP32S3, and in both cases a minimal setup behaves weirdly.

First of there only seems to be around a 1 in 10 chance that zephyr actually "boots" correctly. In most cases it just falls back into a generic exception handler (at least on esp32s3, didn't get debugging working on esp32c3).

In cases where it boots network connectivity is available for a few seconds (responding to pings etc), followed by a fatal exception (on esp32s3) or a load access fault (esp32c3). Both seem to happen inside of the lan865x_interrupt thread.

Building the same project for a rp2040 instead results in a perfectly stable environment, running for hours without any issues.

I've added some debug logging around the code of the lan865x and oa_tc6 driver and found out that in some cases the pointer to the tc6 object suddenly becomes just "4" instead of a normal memory location. The code then tries to use it and that results in the crash. Unfortunately that doesn't seem to be the only place where something like this happens.

There is no custom code running, just spi entries in dts overlay and minimal prj.conf to enable networking and the shell.
I've already tried stripping it down some more by removing my uart changes and disabled logging and the shell, but that doesn't change anything.

Does anybody have some idea on what might be happening here?

[prj.conf.txt](https://github.com/user-attachments/files/16878656/prj.conf.txt)

[xiao_esp32c3.overlay.txt](https://github.com/user-attachments/files/16878665/xiao_esp32c3.overlay.txt)

![2024-09-04_21-42-35](https://github.com/user-attachments/assets/0f9e5304-b01e-4ffd-b0c5-9a60b77942ea)

**Comment by mariopaja** [comment]
Hi @lukas2511,

can you paste your application, overlay & prj.conf as code?
I can test it and give a feedback.

Btw, which zephyr version are you using? 

I have been working with lan856x with zephyr v3.6.0 & v3.7.0 and I didn't have any problems at all.

**Reply by lukas2511** [reply]
Hi @mariopaja

I tried 3.7.0 as well as the main branch at the current latest commit.

The app is just an empty loop:

```
#include <stdio.h>
#include <stdlib.h>
#include <zephyr/kernel.h>

int main(void) {
  while (1) {
    k_msleep(10);
  }
  return 0;
}
```

Initially I had code to blink an LED, but i removed everything to really make sure my code didn't accidentally disturb anything.

prj.conf:
```
# Generic networking options
CONFIG_NETWORKING=y
CONFIG_NET_UDP=y
CONFIG_NET_TCP=y
CONFIG_NET_IPV6=n
CONFIG_NET_IPV4=y
CONFIG_NET_SOCKETS=y
CONFIG_POSIX_API=y
CONFIG_NET_CONNECTION_MANAGER=y

# Kernel options
CONFIG_MAIN_STACK_SIZE=2048
CONFIG_ENTROPY_GENERATOR=y
CONFIG_TEST_RANDOM_GENERATOR=y
CONFIG_INIT_STACKS=y

# Logging
CONFIG_NET_LOG=y
CONFIG_LOG=y
CONFIG_NET_STATISTICS=y
CONFIG_PRINTK=y

# Network buffers
CONFIG_NET_PKT_RX_COUNT=4
CONFIG_NET_PKT_TX_COUNT=4
CONFIG_NET_BUF_RX_COUNT=32
CONFIG_NET_BUF_TX_COUNT=32
CONFIG_NET_CONTEXT_NET_PKT_POOL=y

# IP address options
CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=3
CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=4
CONFIG_NET_MAX_CONTEXTS=10

# Network shell
CONFIG_SERIAL=y
CONFIG_SHELL=y
CONFIG_SHELL_BACKEND_SERIAL=y
CONFIG_NET_SHELL=y

# Network application options and configuration
CONFIG_NET_CONFIG_SETTINGS=y
CONFIG_NET_CONFIG_NEED_IPV4=y
CONFIG_NET_CONFIG_MY_IPV4_ADDR="10.25.11.21"

# Number of socket descriptors might need adjusting
# if there are more than 1 handlers defined.
CONFIG_ZVFS_OPEN_MAX=12

# IP address options
CONFIG_NET_MAX_CONTEXTS=2
CONFIG_NET_MAX_CONN=2

# LAN865x T1S driver specific
CONFIG_NET_L2_ETHERNET=y
# CONFIG_NET_PROMISCUOUS_MODE=y
CONFIG_ETH_LAN865X=y
CONFIG_NET_BUF_DATA_SIZE=64

# Debug options
CONFIG_DEBUG_THREAD_INFO=y
```

overlay:
```
/ {
  aliases {
    led0 = &myled0;
  };

  leds {
    compatible = "gpio-leds";
    myled0: led_0 {
      gpios = <&gpio0 8 GPIO_ACTIVE_LOW>;
    };
  };

  chosen {
    zephyr,console = &uart0;
    zephyr,shell-uart = &uart0;
  };
};

&pinctrl {
	spim2_default: spim2_default {
		group1 {
			pinmux = <SPIM2_MISO_GPIO5>, <SPIM2_SCLK_GPIO4>;
		};
		group2 {
			pinmux = <SPIM2_MOSI_GPIO6>;
			output-low;
		};
	};
};

&spi2 {
	#address-cells = <1>;
	#size-cells = <0>;
	status = "okay";
	pinctrl-0 = <&spim2_default>;
	pinctrl-names = "default";
	cs-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
	eth0: lan865x@0 {
		compatible = "microchip,lan865x";
		reg = <0>;
		rst-gpios = <&gpio0 3 GPIO_ACTIVE_LOW>;
		int-gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
		local-mac-address = [04 05 06 01 02 25];
		plca-enable; /* 1 - PLCA enable, 0 - CSMA/CD enable */
		plca-node-id = <2>; /* PLCA node id range: 0 to 254 */
		plca-node-count = <100>; /* PLCA node count range: 1 to 255 */
		plca-burst-count = <0x0>; /* PLCA burst count range: 0x0 to 0xFF */
		plca-burst-timer = <0x80>; /* PLCA burst timer */
		plca-to-timer = <0x20>; /* PLCA TO value */
		spi-max-frequency = <25000000>;
		status = "okay";
	};
};
```

Thanks in advance

**Reply by mariopaja** [reply]
@lukas2511 did you test eco_client or eco_server sample?

**Reply by lukas2511** [reply]
I tried the echo_server example with the changes from [here](https://github.com/lmajewski/zephyr/commits/t1s-lan8651-nucleo_g431rb-test) (which was linked in the pull-request with the intial implementation), that also doesn't work and after enabling the shell I'm seeing the same error message there.

**Reply by lukas2511** [reply]
Downgrading to 3.6.0 makes bootup a lot more stable, but almost always causes an immediate load access fault:

![2024-09-04_23-18-02](https://github.com/user-attachments/assets/0530f9cb-2250-4cae-b7ee-b7229a49eaef)

**Comment by lukas2511** [comment, marked as answer]
It seems that increasing CONFIG_ETH_LAN865X_IRQ_THREAD_STACK_SIZE fixes the issue... that took me way to long to find...
